homeassistant/components/media_player/epson.py

```diff
@@ -0,0 +1,320 @@
+
+import homeassistant.util as util
+
+DEPENDENCIES = []
+
+KEY_COMMANDS = {
+    "TURN_ON" : [('KEY', '3B')],
+    "TURN_OFF": [('KEY', '3B'), ('KEY', '3B')],
+    "HDMILINK" : [('jsoncallback', 'HDMILINK?')],
+    "CMODE" : [('jsoncallback', 'CMODE?')],
+    "CMODE_CINEMA" : [('CMODE', '15')],
+    "CMODE_NATURAL" : [('CMODE', '07')],
+    "CMODE_BRIGHT" : [('CMODE', '0C')],
+    "CMODE_DYNAMIC" : [('CMODE', '06S')],
+    "VOL_UP" : [('KEY', '56')],
+    "VOL_DOWN" : [('KEY', '57')],
+    "MUTE" : [('KEY', '3E')],
+    "HDMI1" : [('KEY', '4D')],
+    "HDMI2" : [('KEY', '40')],
+    "PC" : [('KEY', '44')],
+    "VIDEO" : [('KEY', '46')],
+    "USB" : [('KEY', '85')],
+    "LAN" : [('KEY', '53')],
+    "WFD" : [('KEY', '56')],
+    "PLAY" : [('KEY', 'D1')],
+    "PAUSE" : [('KEY', 'D3')],
+    "STOP" : [('KEY', 'D2')],
+    "BACK" : [('KEY', 'D4')],
+    "FAST" : [('KEY', 'D5')],
+}
+
+DEFAULT_SOURCES = {
+    'HDMI1' : 'HDMI1',
+    'HDMI2' : 'HDMI2',
+    'PC'    : 'PC',
+    'VIDEO'    : 'VIDEO',
+    'USB'    : 'USB',
+    'LAN'    : 'LAN',
+    'WFD'   : 'WiFi Direct'
+}
+
+INV_SOURCES = {v: k for k, v in DEFAULT_SOURCES.items()}
+
+CMODE_LIST = {
+    '15' : 'Cinema',
+    '07' : 'Natural',
+    '0C' : 'Bright Cinema',
+    '06' : 'Dynamic'
+}
+
+CMODE_LIST_SET = {
+    'cinema': 'CMODE_CINEMA',
+    'natural' : 'CMODE_NATURAL',
+    'bright cinema' : 'CMODE_BRIGHT',
+    'dynamic' : 'CMODE_DYNAMIC'
+}
+
+DATA_EPSON = 'epson'
+DEFAULT_NAME = 'EPSON Projector'
+
+MIN_TIME_BETWEEN_FORCED_SCANS = timedelta(seconds=1)
+MIN_TIME_BETWEEN_SCANS = timedelta(seconds=10)
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Required(CONF_HOST) : cv.string,
+    vol.Optional(CONF_NAME, default=DEFAULT_NAME) : cv.string,
+    vol.Optional(CONF_PORT, default=80) : cv.port,
+    vol.Optional(CONF_SSL, default=False) : cv.boolean
+})
+ATTR_CMODE = 'cmode'
+SUPPORT_CMODE = 33001
+
+SUPPORT_EPSON = SUPPORT_TURN_ON | SUPPORT_TURN_OFF | SUPPORT_SELECT_SOURCE | SUPPORT_CMODE | \
+                SUPPORT_VOLUME_MUTE | SUPPORT_VOLUME_STEP | \
+                 SUPPORT_NEXT_TRACK | SUPPORT_PREVIOUS_TRACK
+                 # SUPPORT_PLAY | SUPPORT_PAUSE | SUPPORT_STOP
+
+
+EPSON_SCHEMA =  MEDIA_PLAYER_SCHEMA.extend({
+    vol.Required(ATTR_ENTITY_ID): cv.entity_id,
+    vol.Optional(ATTR_CMODE): cv.string
+})
+
+# ATTR_TO_PROPERTY = ATTR_TO_PROPERTY.extend(ATTR_CMODE)
+
+_LOGGER = logging.getLogger(__name__)
+
+@asyncio.coroutine
+def async_setup_platform(hass, config, async_add_devices, discovery_info=None):
+
+    if DATA_EPSON not in hass.data:
+        hass.data[DATA_EPSON] = []
+
+    name = config.get(CONF_NAME)
+    host = config.get(CONF_HOST)
+    epson = EpsonProjector(hass, name, host, config.get(CONF_PORT), config.get(CONF_SSL), KEY_COMMANDS)
+
+    epson.update()
+    hass.data[DATA_EPSON].append(epson)
+    async_add_devices([epson], update_before_add=True)
+    # async_add_devices([epson])
```

Comment:
Remove this.

Reply:
do you mean all 4 lines or just commented out line?

I need hass.data[DATA_EPSON] in async_service_handler.